src/test/java/htsjdk/samtools/cram/encoding/CramRecordWriterReaderTest.java

```diff
@@ -0,0 +1,81 @@
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+public class CramRecordWriterReaderTest extends CramRecordTestHelper {
+    private CramCompressionRecord read(final byte[] dataBytes,
+                                       final CompressionHeader header,
+                                       final int refId,
+                                       final Map<Integer, InputStream> inputMap) throws IOException {
+        try (final ByteArrayInputStream is = new ByteArrayInputStream(dataBytes)) {
+            final BitInputStream bis = new DefaultBitInputStream(is);
+
+            final CramRecordReader reader = new CramRecordReader(bis, inputMap, header, refId, ValidationStringency.DEFAULT_STRINGENCY);
+            final CramCompressionRecord recordToRead = new CramCompressionRecord();
+            reader.read(recordToRead);
+            return recordToRead;
+        }
+    }
+
+    private List<CramCompressionRecord> initRTRecords() {
+        final List<CramCompressionRecord> initialRecords = new ArrayList<>();
+
+        // note for future refactoring
+        // createRecord() calls Sam2CramRecordFactory.createCramRecord()
+        // which is the only way to set a record's readFeatures (except via read codec)
+        // which would otherwise be null
+
+        CramCompressionRecord record = createRecord();
+
+        // note for future refactoring
+        // Sam2CramRecordFactory.createCramRecord() sets readBases = qualityScores = new byte[0] if missing
+        // but CramRecordReader.read() sets readBases = qualityScores = null if missing
+        // so we force it here to match the round trips
+
+        record.readBases = record.qualityScores = null;
+
+        initialRecords.add(record);
+        return initialRecords;
+    }
+
+    @Test
+    public void roundTripTest() throws IOException {
+        final List<CramCompressionRecord> initialRecords = initRTRecords();
+
+        // note for future refactoring
+        // createHeader(records) calls CompressionHeaderBuilder.setTagIdDictionary(buildTagIdDictionary(records));
+        // which is the only way to set a record's tagIdsIndex
+        // which would otherwise be null
+
+        final boolean sorted = true;
+        final CompressionHeader header = createHeader(initialRecords, sorted);
+
+        final int refId = Slice.MULTI_REFERENCE;
+        final Map<Integer, ExposedByteArrayOutputStream> outputMap = createOutputMap(header);
+        final byte[] written = write(initialRecords, header, refId, outputMap);
+
+        final Map<Integer, InputStream> inputMap = createInputMap(outputMap);
+        final List<CramCompressionRecord> roundTripRecords = new ArrayList<>(initialRecords.size());
+        for (int i = 0; i < initialRecords.size(); i++) {
+            roundTripRecords.add(read(written, header, refId, inputMap));
+        }
+
+        Assert.assertEquals(roundTripRecords, initialRecords);
```

Comment:
I noticed that `CramCompressionRecord` has defined equals but not hashCode.  I opened https://github.com/samtools/htsjdk/issues/1220 to remember the problem.